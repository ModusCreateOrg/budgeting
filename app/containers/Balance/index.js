// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import type { FormattedAmount } from 'utils/formatAmount';
import { getFormattedBalance, getFormattedInflowBalance, getFormattedOutflowBalance } from 'selectors/transactions';

import BalanceItem from 'components/Balance';
import BalanceRow from 'components/Balance/BalanceRow';

type BalanceProps = {
  balance: FormattedAmount,
  inflow: FormattedAmount,
  outflow: FormattedAmount,
};

class Balance extends React.Component<BalanceProps> {
  render() {
    const { inflow, outflow, balance } = this.props;

    return (
      <BalanceRow>
        <BalanceItem amount={inflow} title="Total Inflow" />
        <BalanceItem amount={outflow} title="Total Outflow" prefix="-" />
        <BalanceItem amount={balance} title="Working Balance" colorize={false} prefix="=" />
      </BalanceRow>
    );
  }
}

const mapStateToProps = state => ({
  balance: getFormattedBalance(state),
  inflow: getFormattedInflowBalance(state),
  outflow: getFormattedOutflowBalance(state),
});

export default connect(mapStateToProps)(Balance);
