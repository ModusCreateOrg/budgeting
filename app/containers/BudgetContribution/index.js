// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom'

import { getTransactionDetails, getInflowBalance, getOutflowBalance } from 'selectors/transactions';
import { roundUp, unsign } from 'utils/arithmetic'

import DonutChart from 'components/DonutChart';
import type { TransactionSummary } from '../../selectors/transactions';


type BudgetContributionProps = {
  transaction: TransactionSummary,
  balance: number
};


function getBudgetPercentageData(transaction: TransactionSummary, totalBalance: number): TransactionSummary[] {
  const trans = { ...transaction }
  trans.value = unsign(trans.value / totalBalance) * 100, 1;
  return [trans].concat([
    {
      category: totalBalance > 0 ? "Inflow Sub Balance" : "Outflow Sub Balance",
      categoryId: Math.ceil(Math.random() * 255) + 255,
      value: unsign((unsign(totalBalance) - unsign(transaction.value)) * 100 / totalBalance)
}]
  );
}

class BudgetContribution extends React.Component<BudgetContributionProps> {

  showAsPercentage = (value) => `${roundUp(value, 2)}%`

  render() {
    const { balance, transaction } = this.props;
    const data = getBudgetPercentageData(transaction, balance);
    return <DonutChart data={data} dataLabel="category" dataKey="category" labelFormatter={this.showAsPercentage} />;
  }
}

const mapStateToProps = (state, { match, location }) => {
  const transactionId = Number(match.params.id);
  const transaction = getTransactionDetails(state, transactionId);
  const balance = transaction.value > 0 ? getInflowBalance(state) : getOutflowBalance(state);
  return {
    transaction, balance
  };
};





export default withRouter(connect(mapStateToProps)(BudgetContribution));
