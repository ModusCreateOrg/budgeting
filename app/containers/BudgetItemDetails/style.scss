@import 'theme/variables';

.budgetItemDetails {

  .goBackButton {
    display: inline-block;
    border: solid 1px $green;
    border-radius: 3px;
    padding: 0.5rem 1.5rem;
    font-size: 16px;
    cursor: pointer;
    box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.1);

    &:hover {
      box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.15);
      opacity: 0.6;
    }

    &:active {
      box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.2);
      opacity: 1;
    }
  }

  .textGreen {
    color: $green;
  }

  .textRed {
    color: $red;
  }

}
