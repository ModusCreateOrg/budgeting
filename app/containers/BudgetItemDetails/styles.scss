.inflowBalance {
  color: #00ff00;
}

.outflowBalance{
  color: #ff0000;
}

.backButton{
  margin-bottom: 1rem;
}
