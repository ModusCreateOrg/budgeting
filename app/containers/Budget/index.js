// @flow
import * as React from 'react';
import transactionReducer from 'modules/transactions';
import categoryReducer from 'modules/categories';
import { injectAsyncReducers } from 'store';
import BudgetGrid from 'containers/BudgetGrid';
import Balance from 'containers/Balance';

// inject reducers that might not have been originally there
injectAsyncReducers({
  transactions: transactionReducer,
  categories: categoryReducer,
});

const BudgetContainer = () => (
  <section>
    <BudgetGrid />
    <Balance />
  </section>
);

export default BudgetContainer;
