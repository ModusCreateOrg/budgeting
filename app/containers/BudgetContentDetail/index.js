// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import formatPercent from 'utils/formatPercent';
import {
  getTransactionById,
  getContributionPercentage,
  getFormattedContributionPercentage,
} from 'selectors/transactions';
import getQueryParam from 'selectors/queries';
import type { Transaction } from 'modules/transactions';
import { withRouter } from 'react-router-dom';
import NavLink from 'components/NavLink';
import DonutChart from 'components/DonutChart';
import BudgetContentSubtitle from 'components/BudgetContentSubtitle';
import styles from './style.scss';

type BudgetContentDetailProps = {
  transaction: Transaction,
  contributionPercentage: number,
  data: [],
};

const BudgetContentDetail = ({ transaction, data, contributionPercentage }: BudgetContentDetailProps) => (
  <section>
    <div className={styles.budgetContentDetail}>
      <h1 className={styles.budgetContentDetailTitle}>{transaction.description}</h1>
      <NavLink to="/budget" label="< Back" styles={styles} />
    </div>
    <BudgetContentSubtitle contributionPercentage={contributionPercentage} />
    <DonutChart data={data} dataLabel="label" dataKey="value" format={value => formatPercent(value).text} />
  </section>
);

const mapStateToProps = (state, ownProps) => {
  const transaction = getTransactionById(state, parseInt(getQueryParam(ownProps.match, 'id'), 10));
  const contributionPercentage = getContributionPercentage(state, transaction);
  return {
    transaction,
    contributionPercentage,
    data: getFormattedContributionPercentage(contributionPercentage, transaction),
  };
};

export default withRouter(connect(mapStateToProps)(BudgetContentDetail));
