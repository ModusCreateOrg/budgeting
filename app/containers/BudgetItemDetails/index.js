// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectAsyncReducers } from 'store';
import { withRouter } from 'react-router-dom';
import type { Match, RouterHistory } from 'react-router-dom';
import { getTransactions, getInflowBalance, getOutflowBalance } from '../../selectors/transactions';
import transactionReducer from '../../modules/transactions';
import type { Transaction } from '../../modules/transactions';
import formatPercentage from '../../utils/formatPercentage';
import formatAmount from '../../utils/formatAmount';
import styles from './style.scss';

// inject reducers that might not have been originally there
injectAsyncReducers({
  transactions: transactionReducer,
});

type BudgetItemDetailsProps = {
  match: Match,
  transactions: Transaction[],
  history: RouterHistory,
  inflowBalance: number,
  outflowBalance: number,
};

export const BudgetItemDetails = ({
  match,
  transactions,
  history,
  inflowBalance,
  outflowBalance,
}: BudgetItemDetailsProps) => {
  const getItem = itemIdToGet => ({ id }) => id === Number(itemIdToGet);
  const { itemId } = match.params;
  const item = transactions.filter(getItem(itemId))[0];

  const isInflow = item.value > 0;
  const percentageOfTotal = formatPercentage(item.value / (isInflow ? inflowBalance : outflowBalance) * 100, {
    decimals: 1,
  });

  return (
    <div className={styles.budgetItemDetails}>
      <div className={`goBackButton ${styles.goBackButton}`} onClick={history.goBack} role="button" tabIndex="0">
        Back
      </div>
      <div>
        <h1>{item.description}</h1>
        <h2>
          Amount {isInflow ? 'in' : 'spent'}: {' '}
          {isInflow ? (
            <span className={styles.textGreen}>{formatAmount(item.value).text}</span>
          ) : (
            <span className={styles.textRed}>{formatAmount(item.value).text}</span>
          )}
        </h2>
        <h3>
          which is {percentageOfTotal} of total {isInflow ? 'inflow' : 'outflow'} {' '}
          {isInflow ? (
            <span className={styles.textGreen}>{formatAmount(inflowBalance).text}</span>
          ) : (
            <span className={styles.textRed}>{formatAmount(outflowBalance).text}</span>
          )}
        </h3>
      </div>
    </div>
  );
};

BudgetItemDetails.propTypes = {
  inflowBalance: PropTypes.number.isRequired,
  outflowBalance: PropTypes.number.isRequired,
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      itemId: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  transactions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    })
  ).isRequired,
};

const mapStateToProps = state => ({
  transactions: getTransactions(state),
  inflowBalance: getInflowBalance(state),
  outflowBalance: getOutflowBalance(state),
});
const mapDispatchToProps = () => ({});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BudgetItemDetails));
