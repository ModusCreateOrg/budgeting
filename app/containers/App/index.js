// @flow
import * as React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';

import ErrorBoundary from 'components/ErrorBoundary';
import AppError from 'components/AppError';
import Header from 'components/Header';
import Budget from 'routes/Budget';
import Reports from 'routes/Reports';
import BudgetDetail from 'containers/BudgetDetail';
import './style.scss';

const App = () => (
  <ErrorBoundary fallbackComponent={AppError}>
    <main>
      <Header />

      <Switch>
        <Route exact path="/budget" component={Budget} />
        <Route path="/reports" component={Reports} />
        <Route path="/budget/:id" component={BudgetDetail} />
        <Redirect exact to="/budget" />
      </Switch>
    </main>
  </ErrorBoundary>
);

export default App;
