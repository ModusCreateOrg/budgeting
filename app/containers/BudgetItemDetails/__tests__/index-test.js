import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { BudgetItemDetails } from 'containers/BudgetItemDetails';

it('renders correctly', () => {
  const mockTransactions = [
    {
      id: 1,
      description: "Trader Joe's food",
      value: -423.34,
      categoryId: 1,
    },
  ];

  const mockCategories = {
    1: 'Groceries',
    2: 'School',
  };

  const mockRouterMatch = {
    path: '/item/:itemId',
    url: '/item/1',
    isExact: true,
    params: { itemId: '1' },
  };

  const mockRouterHistory = {
    goBack: jest.fn(),
  };

  const mockInflowBalance = 3456;
  const mockOutflowBalance = -2346;

  const tree = renderer
    .create(
      <BudgetItemDetails
        match={mockRouterMatch}
        history={mockRouterHistory}
        inflowBalance={mockInflowBalance}
        outflowBalance={mockOutflowBalance}
        transactions={mockTransactions}
        categories={mockCategories}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('correctly calls goBack on Back button click', () => {
  const mockTransactions = [
    {
      id: 1,
      description: "Trader Joe's food",
      value: -423.34,
      categoryId: 1,
    },
  ];

  const mockCategories = {
    1: 'Groceries',
    2: 'School',
  };

  const mockRouterMatch = {
    path: '/item/:itemId',
    url: '/item/1',
    isExact: true,
    params: { itemId: '1' },
  };

  const mockGoBack = jest.fn();

  const mockRouterHistory = {
    goBack: mockGoBack,
  };

  const mockInflowBalance = 3456;
  const mockOutflowBalance = -2346;

  const wrapper = shallow(
    <BudgetItemDetails
      match={mockRouterMatch}
      history={mockRouterHistory}
      inflowBalance={mockInflowBalance}
      outflowBalance={mockOutflowBalance}
      transactions={mockTransactions}
      categories={mockCategories}
    />
  );

  wrapper.find('.goBackButton').simulate('click');

  expect(mockGoBack).toHaveBeenCalled();
});
