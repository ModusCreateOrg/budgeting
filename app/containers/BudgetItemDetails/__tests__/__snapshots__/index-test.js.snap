// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="budgetItemDetails"
>
  <div
    className="goBackButton goBackButton"
    onClick={[Function]}
    role="button"
    tabIndex="0"
  >
    Back
  </div>
  <div>
    <h1>
      Trader Joe's food
    </h1>
    <h2>
      Amount 
      spent
      : 
       
      <span
        className="textRed"
      >
        -$423.34
      </span>
    </h2>
    <h3>
      which is 
      18.0%
       of total 
      outflow
       
       
      <span
        className="textRed"
      >
        -$2,346.00
      </span>
    </h3>
  </div>
</div>
`;
